[DA] 응모화면 API 구현

kimkyunghun3 opened the pull request:
# 개요
- 🔗  이슈링크 : #94 

## 작업 내용
응모 화면 API 구현
- 메인 화면에서 응모 관련 뷰 눌렀을 시 응모 detailView 화면에 detail API GET 조회 구현
- 응모 버튼 눌렀을 시 응모하는 sprinkleApplication API POST 구현

Timer에서 남은시간에 따라 내보내는 곳은 추후 구현하겠습니다!

## 고민사항
ApplyViewModel에서 detailData에서 BehaviorRelay로 했는데 여기에서 초기값을 넣어주는 것을 아무거나 넣어두었습니다.
무언가 불필요한 초기값인건가 싶기도하는데.. 다른 방식을 생각해보아야 할까요?

### 스크린샷
![응모화면](https://user-images.githubusercontent.com/52434820/183839555-e4927f01-73e1-4b88-ac7d-043784793bc8.gif)

kimkyunghun3 committed:
:recycle: 사용하지 않는 파일 삭제

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:recycle: 사용하지 않는 파일 삭제

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: CouponEntity 구현

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: CouponEntity 구현

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: toastView 실패 케이스 추가

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: toastView 실패 케이스 추가

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
Merge remote-tracking branch 'origin/feature/kh/ApplyViewAPI' into feature/kh/ApplyViewAPI

kimkyunghun3 committed:
:bulb: CouponDatum 추가

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: 응모 detail, 뿌리기 응모 API 설정

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: gifticonService 응모 디테일, 뿌리기 응모 네트워크 호출 메서드 구현

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: 응모 디테일 정보 받아올 수 있는 메서드 구현

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: 응모 디테일 정보 받아오는메서드 구현

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:recycle: 응모자수 받는 타입 변경

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:recycle: 내부 디테일 받는 값들 로직 변경

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: 응모 디테일 뷰 detail API, 뿌리기 응모 API ViewModel 구현

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:recycle: MainViewModel에서 응모 뷰모델 초기화 설정 변경

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:bulb: 응모 디테일, 뿌리기 응모 API 로직에 따른 ViewController 구현

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:recycle: 불필요한 코드 제거

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

szzang-kr commented on Projects/App/Sources/Apply/ApplyViewController.swift:
```diff
@@ -13,25 +13,79 @@ import RxSwift
-    private let applyButton = DDIPCTAButton()
-    
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+    }
```
사용하지 않는 내용이라면 없어도 될 것 같습니다 😃

kimkyunghun3 replied:
제거했습니다!

szzang-kr commented on Projects/App/Sources/Apply/ApplyViewController.swift:
```diff
@@ -13,25 +13,79 @@ import RxSwift
+        applyButton.setTitle(title: "응모 마감")
+    }
+
+    override func bind() {
+        applyButton.rx.tap.asObservable()
```
override라면 super를 호출해주는것 잊지마세요 !
```suggestion
    override func bind() {
        super.bind()
        applyButton.rx.tap.asObservable()
```

kimkyunghun3 replied:
앗.. override method일거라고 인지못했었네요 감사합니다!

szzang-kr commented on Projects/App/Sources/Apply/ApplyViewController.swift:
```diff
@@ -13,25 +13,79 @@ import RxSwift
+            })
+            .disposed(by: disposeBag)
+
+        _ = viewModel.showToastView
+            .bind { state in
+            if state == true {
+                self.setApplyCompletButtonState()
+                self.showSuccessToast(category: self.viewModel.detailData.value.imageName)
+                self.toastView.showToastView(with: self.view)
+            } else {
+                self.setFinishButtonState()
+                self.showFailToast()
+                self.toastView.showToastView(with: self.view)
+            }
+        }
+        .disposed(by: disposeBag)
```
1. _ = 로 사용하는 이유가 있을까요?

2. bind의 내용은 closure이기 때문에 self를 strong reference하게되면 reference cycle이 발생할 수 있습니다. [요기](http://csl.cool/2021/09/22/ios-dev/swift/memory-leak-in-rx-swift/#bind-single-to-bloc) 참고 해보시면 좋을 것 같습니다.

kimkyunghun3 replied:
1. 처음에 설게하는 도중에 반환값이 있는것처럼 나와서 이를 썼었는데 지금 확인해보니 불필요한 거 같네요!!! 그래서 제거했습니다.!
2. 아 이부분 까먹지 않고 해야 메모리 누수를 막을 수 있는데 자꾸 까먹네요!! 다음부터 잘 인지해서 쓰겠습니다!! 참고 감사합니다 ㅎㅎ

szzang-kr commented on Projects/App/Sources/Apply/ApplyViewController.swift:
```diff
@@ -13,25 +13,79 @@ import RxSwift
+        }
+        .disposed(by: disposeBag)
+
+        _ = viewModel.detailData.subscribe(onNext: { entity in
+            self.applyGifticonView.setParticipant(participants: entity.participants)
+            self.applyGifticonView.setBrand(name: entity.brandName)
+            self.applyGifticonView.setCategory(name: entity.category)
+            self.applyGifticonView.setProductName(name: entity.merchandiseName)
+            self.applyGifticonView.setExpirationDate(name: entity.expiredAt)
+            self.applyGifticonView.setImageIcon(imageName: entity.imageName)
+
+            // TODO: 형변환 및 로직 필요
+//            self.applyGifticonView.setCountdownDate(date: entity.sprinkleAt.)
+        })
+        .disposed(by: disposeBag)
```
마찬가지로 self 의 strong reference가 문제가 될 것 같습니다 😓

kimkyunghun3 replied:
모두 약한참조 추가했습니다..!

szzang-kr commented on Projects/App/Sources/Apply/ViewModel/ApplyViewModel.swift:
```diff
@@ -7,15 +7,48 @@
 //
 
 import Foundation
+import RxSwift
+import RxCocoa
```
```suggestion
import Foundation

import RxCocoa
import RxSwift
```

1. 현재까지 사용했던 컨벤션을 생각해주세요 :)

2. RxCocoa를 왜 import하게 되었나요?

kimkyunghun3 replied:
1. 바로수정했습니다..! 
2. 처음에 RxCocoa 관련된 것을 사용하다보니 이것을 사용했는데 생각해보니 이에 대한 것을 쓰고있지않네요.! 로직변경하며 교체하지못했었던거 같습니다. 하지만 RxRelay가 RxCocoa 내부에 있어서 에러가 나지않아 인지 못했던거같습니다..!

szzang-kr commented on Projects/App/Sources/Apply/ViewModel/ApplyViewModel.swift:
```diff
@@ -7,15 +7,48 @@
 protocol ApplyViewModelProtocol {
-    var gifticonId: Int { get }
+    var showToastView: PublishRelay<Bool> { get }
+    var detailData: BehaviorRelay<CouponDatum> { get }
```
말씀해주신 부분이 요기인것 같은데, 초기데이터를 넣기 어렵다면 아래와같이 사용할 수 있습니다!

`var detailData: BehaviorRelay<CouponDatum?> { get } `

단, 해당 relay에서 방출하는 이벤트는 CouponDatum? 타입이기 때문에 옵셔널 바인딩이 필요해집니다. :)

kimkyunghun3 replied:
요렇게 하는게 더 깔끔한거같네요! 옵셔널로 주고 초기값을 nil로 하는 방식으로 수정했습니다.
그에 따른 옵셔널 바인딩을 하긴 했는데 이부분에 대한 에러처리는 고민해보아야할거같습니다!

szzang-kr commented on Projects/App/Sources/Apply/ViewModel/ApplyViewModel.swift:
```diff
@@ -7,15 +7,48 @@
+    private let gifticonId: Int
     
-    init(gifticonId: Int) {
+    init(gifticonId: Int, network: Networking) {
         self.gifticonId = gifticonId
+        self.gifticonService = GifticonService(network: network)
```
👍

szzang-kr commented on Projects/App/Sources/Apply/ViewModel/ApplyViewModel.swift:
```diff
@@ -7,15 +7,48 @@
+                guard let responseData = responseModel.data else { return
+                }
+                self?.detailData.accept(responseData)
+            }, onFailure: { error in
+            })
+            .disposed(by: disposeBag)
```
에러처리가 필요한 부분이라면 로직이 필요할것 같습니다!

아직 에러케이스에대한 설계가 되지 않았다면, 사용하셨던 `// TODO` 를 통해 잊어버리지 않도록 마킹해두는것도 좋은 방법입니다. :)

kimkyunghun3 replied:
음 이부분에서 에러처리에 대한 로직에 대해 아직 이야기 된 기획이 없어서 딱히 넣어두지않았고, 그로인해 남겨둔 것이 아니였기에 일단 제거했습니다.

추후에 에러처리가 필요하다면 추가하도록 하겠습니다!

szzang-kr commented on Projects/App/Sources/Driver/Service/GifticonAPI.swift:
```diff
@@ -25,15 +27,10 @@ extension GifticonAPI: NetworkRequestable {
-        case .registerSprinkle:
-            return .post
+        case .sprinkleApplication(let id):
+            return "/api/v1/sprinkle/\(3)/apply"
+        case .sprinkleDetail(let id):
+            return "/api/v1/sprinkle-info/\(3)"
```
해당부분은 임시로 3을 지정해두신걸까요?

실제 API가 배포된 후에도 고정값으로 API를 호출할 수 있으니 `// TODO`주석을 추가하여 잊어버리지 않도록 주의하세요 !! 😮

kimkyunghun3 replied:
앗! 일단 id로 바꾸어두었습니다! 임시로 3으로 해둔것이 맞습니다!! 까먹지않겠습니다 ㅎㅎ

szzang-kr commented on Projects/App/Sources/Driver/Service/GifticonAPI.swift:
```diff
@@ -43,6 +40,10 @@ extension GifticonAPI: NetworkRequestable {
             return nil
         case let .categoryList(model):
             return model
+        case .sprinkleApplication(_):
+            return nil
+        case .sprinkleDetail(_):
+            return nil
```
```suggestion
        case .sprinkleApplication:
            return nil
        case .sprinkleDetail:
            return nil
```

사용하지 않는 associated value의 경우 위와같이 사용할 수 있습니다 !!

또한 case의 return이 같다면 아래와 같이 작성할수도 있습니다 😄 
```suggestion
        case .sprinkleApplication, .sprinkleDetail:
            return nil
```

kimkyunghun3 replied:
이와 관련된 스위치문을 모두 수정했습니다! 동일한 것은 묶어서 하니 더 가독성이 높아지는거 같네요 감사합니다!

szzang-kr commented on Projects/App/Sources/Driver/Service/GifticonAPI.swift:
```diff
@@ -54,6 +55,25 @@ extension GifticonAPI: NetworkRequestable {
             return .default
         case .registerSprinkle:
             return .multipartHeader
+        case .sprinkleApplication(_):
+            return .default
+        case .sprinkleDetail(_):
+            return .default
+        }
+    }
+
+    var method: HTTPMethod {
+        switch self {
+        case .categoryList(_):
+            return .get
+        case .sprinkleApplication(_):
+            return .post
+        case .sprinkleDetail(_):
+            return .get
+        case .categories:
+            return .get
+        case .registerSprinkle(_):
+            return .post
```
[이전 코멘트](https://github.com/mash-up-kr/GGiriGGiri_iOS/pull/118#discussion_r942133275)의 내용을 참고하시면 좋을것 같아요!

kimkyunghun3 replied:
이와 관련된 스위치문을 모두 수정했습니다! 동일한 것은 묶어서 하니 더 가독성이 높아지는거 같네요 감사합니다!

szzang-kr commented on Projects/App/Sources/Usecase/CouponEntity.swift:
```diff
@@ -8,6 +8,44 @@
+        case brandName, merchandiseName, category, expiredAt, participants, sprinkleAt, participateIn
+    }
+
+    var imageName: DDIPAsset.name {
+        switch category {
+        case "CAFE":
+            return .iconCafedesert
+        case "DELIVERY":
+            return .iconChicken
+        case "ICECREAM":
+            return .iconIcecream
+        case "CONVENIENCE_STORE":
+            return .iconDrugStore
+        case "FAST_FOOD":
+            return .iconFastfood
+        case "VOUCHER":
+            return .iconCoupon
+        case "ETC":
+            return .iconGuitar
+        default:
+            return .iconLogoCharacter
+        }
```
카테고리 API 리팩터링 적용 후 함께 변경해볼만한 내용이네요 !! 기대해주세요 🤭 

(cc. @ios-h)

szzang-kr commented on Projects/DesignSystem/Sources/Component/ToastView/DDIPToastView.swift:
```diff
@@ -22,6 +22,14 @@ public class DDIPToastView: UIView, AddViewsable {
                 return ("등록 실패", "다시 한 번 시도해주세요.")
             case .apply:
                 return ("응모 완료", "응모가 완료되었어요!\n 응모 결과는 마감 후에 확인할 수 있어요.")
+            case .applyMyGifticonFail:
+                return ("응모 실패", "내가 등록한 기프티콘에는\n 응모할 수 없어요!")
+            case .applyFail:
+                return ("응모 실패", "다시 한번 시도 해주세요.")
+            case .saveFail:
+                return ("저장 실패", "다시 한번 시도해주세요.")
+            case .registerFail:
+                return ("등록 실패", "다시 한번 시도해주세요.")
```
서버에서 받아오는 message 케이스가 다양해지면 해질수록 케이스가 늘어날 것 같습니다!

또한 앱이 배포되고, 서버에서 message가 추가된다면 앱은 case를 추가하는 앱 배포가 이루어져야 합니다.

이런 상황을 고려하여 어떻게 리팩터링 하면 좋을지 고민해봅시다. 🌝 @ios-h  @kimkyunghun3

kimkyunghun3 replied:
계속 고민해보았는데 구현부가 이곳 하나라서 이곳에서만 바꾸어주면 된다고 생각했습니다.
그러면 다른 곳에서 코드 중복이 없을 것이고 변경되는 곳은 이곳이라 괜찮은 방법이라고 생각은 했지만, 이렇게 case가 늘어남에 따라 조금 길어지는 코드가 보기 불편했던건 사실입니다..!

그래서 어떻게 리팩을 해야할까 고민해보았는데... 프로토콜을 사용하거나.. enum내부에서 메서드를 만들어서 하거나 등을 생각해보았는데
결국 비슷하게 코드가 이루어질거같다는 생각이 들었습니다.

케이스를 분리하기에는 하나로 처리할 수 있는 편리함을 없애야하고 그래서 정확하게 어떤 리팩이 필요할지 방향을 못잡겠습니다.
힌트를 주시면 조금 더 찾아보겠습니다!

szzang-kr reviewed (changes requested):
수고하셨습니다!! QA까지 호로록 끝내봅시다 💪

kimkyunghun3 committed:
:recycle: 불필요한 코드 제거 및 강한참조사이클 문제 해결

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:recycle: API 타입에 따른 설정 변경

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:recycle: detailData 초기값 없도록 구현에 따른 로직 수정

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:recycle: 불필요한 코드 제거

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

szzang-kr commented on Projects/App/Sources/Apply/ApplyViewController.swift:
```diff
@@ -13,25 +13,86 @@ import RxSwift
+
+        viewModel.detailData
+            .subscribe(onNext: { [weak self] entity in
+            guard let participants = entity?.participants,
+                  let brandName = entity?.brandName,
+                  let category = entity?.category,
+                  let merchandiseName = entity?.merchandiseName,
+                  let expiredAt = entity?.expiredAt,
+                  let imageName = entity?.imageName
+            else { return }
```
guard let을 모두 만족해야 뷰에 정보를 뿌려주는 구조인데, 예를들어 다른값은 모두 정상인데 brandName 만 이상한 경우에는 모두 표시 되지 않을것 같습니다.

guard let 대신 Nil coalescing(??) 을 사용해보면 어떨까요?

kimkyunghun3 replied:
아하! 그렇게 말씀하시니 그게 더 나을거같네요! 수정했습니다!

kimkyunghun3 committed:
:recycle: 불필요한 코드 제거

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
:recycle: 옵셔널 바인딩 방법 변경

# :tada: = 프로젝트 시작 (#)
# :sparkles: = create 파일 생성 시 (#)
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시 (#)
# :rocket: = release 배포 시 (#)
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우) (#)
# :memo: = 문서 작성 시 (#)
# :art: = 디자인 시스템 관련 내용 (Color / Asset) (#)
# :hammer: = 테스트 코드 작성 (#)
################

# 커밋 타입 : 커밋 제목 (계정명/저장소명#이슈번호) 형식으로 작성하며 제목은 최대 50글자 정도로만 입력
# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

kimkyunghun3 committed:
Merge branch 'develop' into feature/kh/ApplyViewAPI
